### DISCUSSION TITLE
clangd over tramp - multiple instances, difficulty launching

### DISCUSSION THREAD

**Discussion by mikezackles** [joaotavora/eglot#1050] [Q&A] [answered]
I'm working on a large codebase over tramp using doom emacs with eglot 1.8 and emacs 28. (Sorry, I'm asking here because I lack the emacs/elisp knowledge to debug this better myself.) I've not made any eglot customizations myself beyond setting the clangd command via `eglot-server-programs`.

I've been using projectile to start dired in the project root, and I'll navigate to a c++ source file in a subdirectory from there. From there, two problematic things are happening:

1) The clangd server is not reliably launched. Often eglot will display a message that it is waiting and time out, but I never see the executable running on the machine. But sometimes it works immediately, and I've noticed I'm usually (always?) successful if I launch emacs after launching an instance of visual studio code (which launches its own clangd instance).
2) After connecting things work pretty well, but I'm noticing multiple clangd instances getting spun up. It seems like every time I navigate to another folder in the project, another clangd process is launched, which means I end up accumulating quite a number of clangd instances by the end of my day. Each clangd has an associated eglot message buffer. Perhaps there are some issues identifying the project root?

I would appreciate any pointers on what might be wrong or how to better debug this myself. I know a vanilla install might be better than doom, but I haven't yet had time to devote to exploring that.

**Comment by joaotavora** [comment, marked as answer]
Problem number 2 is possibly related to your use of projectile.  You must make sure that, all Eglot-ness aside, the function `project-current` returns the same value when invoked on all of those directories. You can check this with `M-: (project-current) RET`.  If that doesn't give you the same value, then you must setup your `project.el` setup so that it does.  Normally  it isn't a problem when a `.git` directory is dominating the whole project, which is true for a large percentage of cases. @dgutov is the `project.el` specialist.

As to problem number 1, it's hard to say without a controlled experiment that you can share with us, but you could try fiddling with `eglot-sync-connect` which is the number of seconds that Eglot will wait for a synchronous, GUI-blocking connection attempt with the server before waiting in the background for this connection. 

> I'm usually (always?) successful if I launch emacs after launching an instance of visual studio code (which launches its own clangd instance).

Maybe by then `clangd` will be faster to load because some of its shared libraries are in the memory cache, or because it is sharing some cached information done upfront.

**Reply by mikezackles** [reply]
Thanks so much for the prompt and thoughtful replies.

Increasing the value for `eglot-connect-timeout` did seem to more or less fix problem 1. I guess I was confused because sometimes it seems to connect right away, and sometimes it takes a bit.

Indeed, `(project-current)` returns nil, so I will continue looking into setting that up. For the record, there is a single `.git` directory at the top level of the project.

**Reply by mikezackles** [reply]
The project root issue was caused by doom modifying `vc-ignore-dir-regexp`: https://github.com/doomemacs/doomemacs/blob/7fd6cd4398ad78504bfff46791a5a47a9520b0a6/lisp/doom-editor.el#L420-L427

After commenting that code, the project gets detected successfully.

**Reply by mikezackles** [reply]
Bug filed for completeness: https://github.com/doomemacs/doomemacs/issues/6834

**Comment by joaotavora** [comment]
This is yet another TRAMP-related problem, it appears.  Eglot frequently gets reports of problematic TRAMP situations. And mixed reports of problems that went away, and came back.  

We badly need someone with a working TRAMP setup that investigates these problems systematically and meticulously.  I haven't used TRAMP in a long time: I just don't need it (I remember having problems with it even before Eglot).
